cpp/src/arrow/csv/reader.cc

```diff
@@ -970,6 +972,181 @@ class StreamingReaderImpl : public ReaderMixin,
   std::shared_ptr<std::atomic<int64_t>> bytes_decoded_;
 };
 
+
+
+// this is just like a MapGenerator but the map fun returns a thing instead of a future
+template <typename T, typename ApplyFn,
+          typename Applied = arrow::detail::result_of_t<ApplyFn(const T&)>,
+          typename V = typename EnsureResult<Applied>::type::ValueType>
+AsyncGenerator<V> MakeApplyGenerator(AsyncGenerator<T> source_gen, ApplyFn apply_fun, internal::Executor* cpu_exec) {
```

Comment:
I think this should actually go in `async_generator.h`.  I think I told you otherwise once before but I was thinking it was going to be more tailored to the CSV reader.  This seems pretty generic.  Also, it should ideally be independently unit tested with a few tests.